### PR TITLE
Support returning `Result<OpaqueRust, OpaqueRust>` from async function

### PR DESCRIPTION
**Pull request by NiwakaDev**
This PR addresses #157 and adds support for returning a `Result<OpaqueRust, OpaqueRust>` from async functions.

Here's an example of using this. 

```rust
//Rust
#[swift_bridge::bridge]
mod ffi {
    extern "Rust" {
        type OpaqueType1;
        type OpaqueType2;
        async fn some_function() -> Result<OpaqueType1, OpaqueType2>;
    }
}
```

```Swift
//Swift
do {
    let opaqueType1 = try await some_function()
} catch {
    //...
}
```

### CODE REVIEW

File: crates/swift-bridge-ir/src/bridged_type.rs [hunk 1]
```diff
@@ -54,6 +54,13 @@ pub(crate) trait BridgeableType: Debug {
     /// Whether or not this is a `Result<T,E>`.
     fn is_result(&self) -> bool;
 
+    /// Extract the Swift representation of each variant from Rust's Result<T,E>.
```
Comment (chinedufn) [inline crates/swift-bridge-ir/src/bridged_type.rs:57]: This exact Swift results variant feels a bit too specific to belong in the `BridgeableType` trait.

---

Here's what I'm thinking instead:

---

We add an `as_result` method to the `BridgeableType` trait

```rust
trait BridgeableType {
    // ...

    fn is_result(&self) -> bool {
        self.as_result().is_some()
    }

    fn as_result(&self) -> Option<&BuiltInResult>;
}
```

---

Then we can use `.as_result()` to get back the BuiltInResult, and use a method(s) on the built in result to generate whatever information we need when generating the async support code on the Swift side.

**Review by chinedufn — COMMENTED** [review]
@NiwakaDev this is truly outstanding work. AWESOME.

I left some minor feedback, but overall this is excellent. Really impressive stuff. Thanks a lot.

File: book/src/bridge-module/async-functions/README.md [hunk 2]
```diff
@@ -18,3 +18,19 @@ async fn some_async_function(list: Vec<u8>) -> ffi::MyStruct {
     ffi::MyStruct
 }
 ```
+
+## Return `Result<OpaqueType, OpaqueType>` from async functions
```
[the remark this answers is not in the document]
Comment (chinedufn) [reply book/src/bridge-module/async-functions/README.md:22]: As well as a Swift block below it:

```swift
// Swift

let totalUsers = await user_count()

do {
    let user = try await load_user("https://example.com/users/5")
} catch let error as ApiError {
    // ... error handling ...
}
```

File: crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs [hunk 3]
```diff
@@ -533,3 +533,107 @@ void __swift_bridge__$SomeType$some_method(void* callback_wrapper, void __swift_
         .test();
     }
 }
+
+mod extern_rust_async_function_returns_result_opaque {
```
Comment (chinedufn) [inline crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs:537]: Mind adding a `Verify that ...` comment like the other bridge modules?

**Comment by chinedufn** [comment]
Just wanted to say again that this is a really great PR. Thanks for implementing this!!

File: crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs [hunk 4]
```diff
@@ -533,3 +533,107 @@ void __swift_bridge__$SomeType$some_method(void* callback_wrapper, void __swift_
...
+    func onComplete(cbWrapperPtr: UnsafeMutableRawPointer?, rustFnRetVal: __private__ResultPtrAndPtr) {
+        let wrapper = Unmanaged<CbWrapper$some_function>.fromOpaque(cbWrapperPtr!).takeRetainedValue()
+        if rustFnRetVal.is_ok {
+            wrapper.cb(.success(SomeType1(ptr: rustFnRetVal.ok_or_err!)))
+        } else {
+            wrapper.cb(.failure(SomeType2(ptr: rustFnRetVal.ok_or_err!)))
+        }
+    }
+
+    return try await withCheckedThrowingContinuation({ (continuation: CheckedContinuation<SomeType1, Error>) in
```
[the remark this answers is not in the document]
Comment (NiwakaDev) [reply crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs:599]: I tried to do this, I got the following error.

```
Testing failed:
	Cannot convert value of type '(CheckedContinuation<AsyncResultOpaqueRustType1, AsyncResultOpaqueRustType2>) -> Void' to expected argument type '(CheckedContinuation<AsyncResultOpaqueRustType1, any Error>) -> Void'
	Testing cancelled because the build failed.

```

File: crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs [hunk 5]
```diff
@@ -533,3 +533,107 @@ void __swift_bridge__$SomeType$some_method(void* callback_wrapper, void __swift_
...
+
+mod extern_rust_async_function_returns_result_opaque {
+    use super::*;
+
+    fn bridge_module() -> TokenStream {
+        quote! {
+            #[swift_bridge::bridge]
+            mod ffi {
+                extern "Rust" {
+                    type SomeType1;
```
[the remark this answers is not in the document]
Comment (NiwakaDev) [reply crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs:545]: Sounds good!

File: crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (NiwakaDev) [reply crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs:599]: I made the minimal code to reproduce this error.

```Swift
struct ZeroError: Error{}

func hoge(value: Int) async throws -> Int {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Int, ZeroError>) in
        if value == 0{
            continuation.resume(throwing: ZeroError())
        }else{
            continuation.resume(returning: value*2)
        }
    }
}
```

[the remark this answers is not in the document]
Comment (chinedufn) [reply crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs:599]: Thanks for the minimal example.

This compiles:

```swift
struct ZeroError: Error{}

func hoge(value: Int) async throws -> Int {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Int, _>) in
        if value == 0{
            continuation.resume(throwing: ZeroError())
        }else{
            continuation.resume(returning: value*2)
        }
    }
}
```

Since `ZeroError` doesn't compile leaving it as `Error` is fine.

Would just leave a comment somewhere in the test.

Maybe:

```
TODO: Replace `Error` with the concrete error type `SomeType2`. 
 As of Feb 2023 using the concrete error type leads to a compile time error.
 This seems like a bug in the Swift compiler.
```

File: SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/AsyncTests.swift [hunk 6]
```diff
@@ -41,6 +41,25 @@ class AsyncTests: XCTestCase {
...
 
     
+    /// Verify that we can pass and return a Result<OpaqueRust, OpaqueRust> to an async Rust function
+    func testSwiftCallsRustAsyncFnReflectResultOpaqueRust() async throws {
+        
+        //Should return an AsyncResultOpaqueRustType1 type.
+        do {
+            let result = try await rust_async_func_reflect_result_opaque_rust(.Ok(AsyncResultOpaqueRustType1(10)))
+        } catch {
+            XCTFail()
```
[the remark this answers is not in the document]
Comment (NiwakaDev) [reply SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/AsyncTests.swift:51]: What about also replacing `fatalError()` with `XCTFail()` in other tests?

**Comment by NiwakaDev** [comment]
Addressed your review! Please let me know if there are any omissions.

File: SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/AsyncTests.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (chinedufn) [reply SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/AsyncTests.swift:51]: Great idea, please go for it!

File: crates/swift-bridge-ir/src/codegen/generate_swift/generate_function_swift_calls_rust.rs [hunk 7]
```diff
@@ -194,33 +194,89 @@ pub(super) fn gen_func_swift_calls_rust(
         let func_ret_ty = function.return_ty_built_in(types).unwrap();
         let rust_fn_ret_ty =
             func_ret_ty.to_swift_type(TypePosition::FnReturn(HostLang::Rust), types);
-
         let (maybe_on_complete_sig_ret_val, on_complete_ret_val) = if func_ret_ty.is_null() {
             ("".to_string(), "()".to_string())
         } else {
-            (
-                format!(
```
Comment (chinedufn) [inline crates/swift-bridge-ir/src/codegen/generate_swift/generate_function_swift_calls_rust.rs:202]: We can undo this change and go back to the original code.

The `if` and `else` blocks that were added are identical.

**Review by chinedufn — COMMENTED** [review]
One minor bit of cleanup than looks good to me!

File: crates/swift-bridge-ir/src/codegen/generate_swift/generate_function_swift_calls_rust.rs (hunk 7 above)
Comment (NiwakaDev) [reply crates/swift-bridge-ir/src/codegen/generate_swift/generate_function_swift_calls_rust.rs]: Done.
